### PR TITLE
Feature#2

### PR DESCRIPTION
**Pull request by baeyuyeon**
## 개요
member controller 위주로 수정하였습니다.

## 주요 변경 사항
1. aplication.properties 의 설정을 통한 request uri변경
```
#uri prefix for all controller
server.servlet.context-path=/board/v1
````  

2. Controller에서 RequestMappingmethod 를 통한 CRUD구현
(MemberController.java 소스 참고)
```
  GET : Read
  POST : Create
  PUT : Update
  DELETE :  Delete
```

3. ExceptionHandler를 통한 오류 코드 리턴
(MemberController.java 소스 참고)

## 특이 사항
1. 조회시에 POST 로 받을 수 있는 방법이 궁금합니다.
2.  @ExceptionHandler 처리가 Controller에 있어도 되는지 궁금합니다.

### CODE REVIEW

File: src/main/java/com/board/toyproject/controller/MemberController.java [hunk 1]
```diff
@@ -3,107 +3,126 @@
...
+
+
+    //---------------------------------------------------------//
     /**
      * 회원가입
      * @param member
      * @return
      */
-    @RequestMapping(value = "/create_member", method = {RequestMethod.GET, RequestMethod.POST})
+    @RequestMapping(value = "/members", method = {RequestMethod.POST})
```
Comment (f-lab-dori) [inline src/main/java/com/board/toyproject/controller/MemberController.java:53]: @RequestMapping을 class에 선언해서 하면 아래도 다 적용이 됩니다.

File: src/main/java/com/board/toyproject/controller/MemberController.java [hunk 2]
```diff
@@ -3,107 +3,126 @@
...
     }
 
+
     /**
      * 멤버 정보 수정
      * @param member
      * @return
      */
-    @RequestMapping(value = "/updateMember", method = {RequestMethod.GET, RequestMethod.POST})
+    @RequestMapping(value = "/members", method = {RequestMethod.PUT})
```
Comment (f-lab-dori) [inline src/main/java/com/board/toyproject/controller/MemberController.java:94]: 하나라서 {RequestMethod.PUT} -> RequestMethod.PUT로 하셔도 될 듯요?

File: src/main/java/com/board/toyproject/controller/MemberController.java [hunk 3]
```diff
@@ -3,107 +3,126 @@
...
+        //Map<String, Object> resultMap = new HashMap<>();
         if (member == null) {
-            resultMap.put("result", "멤버 정보를 입력하세요.");
+            throw new NoSuchElementException("수정하려는 멤버정보를 입력해주세요.");
         }else{
+            if(!memberService.findByMemberId(member.getMemberId()).isPresent()){
+                throw new NoSuchElementException("수정하려는 아이디가 존재하지 않습니다.");
+            }
+
             String memberId=memberService.updateMember(member);
```
Comment (f-lab-dori) [inline src/main/java/com/board/toyproject/controller/MemberController.java:105]: =의 앞과 뒤는 스페이스를 꼭 넣어주세요 모든 코드 컨벤션은 일관성있게 해주세요

File: src/main/java/com/board/toyproject/controller/MemberController.java [hunk 4]
```diff
@@ -3,107 +3,126 @@
...
      * @return
      */
-    @RequestMapping(value = "/deleteMember", method = {RequestMethod.GET, RequestMethod.POST})
+    @RequestMapping(value = "/members", method = {RequestMethod.DELETE})
     @ResponseBody
-    public Map<String, Object> deleteMember(@RequestBody Member member) {
-        System.out.println("member = " + member);
+    public String deleteMember(@RequestBody Member member) {
+        if (member == null) {
+            throw new NoSuchElementException("수정하려는 멤버정보를 입력해주세요.");
```
Comment (f-lab-dori) [inline src/main/java/com/board/toyproject/controller/MemberController.java:119]: 이 경우는 400 에러 아닐까요?

File: src/main/java/com/board/toyproject/service/MemberService.java [hunk 5]
```diff
@@ -2,9 +2,12 @@
...
 import com.board.toyproject.repository.MemberRepository;
+import com.mysql.jdbc.exceptions.jdbc4.MySQLIntegrityConstraintViolationException;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.DuplicateKeyException;
 
 import java.util.List;
+import java.util.NoSuchElementException;
 import java.util.Optional;
 
 public class MemberService {
```
Comment (f-lab-dori) [inline src/main/java/com/board/toyproject/service/MemberService.java:13]: interface를 만들어 느슨한 결합을 만들어주세요

**Review by raundev — APPROVED** [review]

File: src/main/java/com/board/toyproject/controller/MemberController.java [hunk 6]
```diff
@@ -3,107 +3,126 @@
...
-    @GetMapping("/requestMemberById")
+    @RequestMapping(value = "/members", method = {RequestMethod.GET})
     @ResponseBody
-    public Member requestMemberById(@RequestParam(value = "memberId",required = false) String memberId){
-        Member member = memberService.findByMemberId(memberId).orElse(null);
-        System.out.println("member = " + member);
-        return member;
-    }
+    public List<Member> requestMemberById(@RequestParam(value = "memberId",required = false) String memberId
+                                    ,@RequestParam(value = "name",required = false) String name){
```
[the remark this answers is not in the document]
Comment (baeyuyeon) [reply src/main/java/com/board/toyproject/controller/MemberController.java:73]: @f-lab-dori 멘토님~!
혹시 'required = false' 를 붙일 필요가 없다는 말씀이신가요??
required =true가 default여서 false를 지우면 안될 것 같습니다!

File: src/main/java/com/board/toyproject/controller/MemberController.java [hunk 7]
```diff
@@ -3,107 +3,126 @@
...
         }
-        return resultMap;
     }
     /**
      * 멤버 탈퇴
      * @param member
      * @return
      */
-    @RequestMapping(value = "/deleteMember", method = {RequestMethod.GET, RequestMethod.POST})
+    @RequestMapping(value = "/members", method = {RequestMethod.DELETE})
```
[the remark this answers is not in the document]
Comment (baeyuyeon) [reply src/main/java/com/board/toyproject/controller/MemberController.java:115]: @f-lab-dori 멘토님~!

https://medium.com/@nadinCodeHat/rest-api-naming-conventions-and-best-practices-1c4e781eb6a5
위 글에 따르면 
'Use Pluralized Nouns for resources
Use plural when possible unless they are singleton resources.'
라고 되어있어서 모두 복수형으로 통일하였습니다!

File: src/main/java/com/board/toyproject/controller/MemberController.java [hunk 8]
```diff
@@ -1,109 +1,108 @@
...
+import org.springframework.http.ResponseEntity;
 import org.springframework.stereotype.Controller;
 import org.springframework.web.bind.annotation.*;
+import org.springframework.web.context.request.WebRequest;
 
 import java.util.*;
 
-@Controller
+@RestController
+@RequestMapping(value = "/members")
```
Comment (f-lab-dori) [inline src/main/java/com/board/toyproject/controller/MemberController.java:17]: 여기에 /v1/같은 버전 정보를 넣으시는 게 어떠신가요? 그리고 members가 아니고 member가 더 맞지 않을까요?

File: src/main/java/com/board/toyproject/controller/MemberController.java [hunk 9]
```diff
@@ -1,109 +1,108 @@
...
-     * 아이디로 멤버 찾기
+     * 멤버 찾기
      * @param memberId
      * @return
      */
-    @GetMapping("/requestMemberById")
+    @GetMapping
+    //@PostMapping
     @ResponseBody
-    public Member requestMemberById(@RequestParam(value = "memberId",required = false) String memberId){
```
Comment (f-lab-dori) [inline src/main/java/com/board/toyproject/controller/MemberController.java:47]: memberId가 필수이면 path variable로 넣으시는 게 어떠신가요?

File: src/main/java/com/board/toyproject/controller/MemberController.java [hunk 10]
```diff
@@ -1,109 +1,108 @@
...
-        Member member = memberService.findByMemberId(memberId).orElse(null);
-        System.out.println("member = " + member);
-        return member;
-    }
+    public List<Member> requestMemberById(@RequestParam(value = "memberId", required = false) String memberId
+                                    ,@RequestParam(value = "name", required = false) String name){
```
Comment (f-lab-dori) [inline src/main/java/com/board/toyproject/controller/MemberController.java:54]: 이름으로 찾는거랑 memberId로 찾는거랑 메서드를 분리하는 것이 낫을 듯요

File: src/main/java/com/board/toyproject/controller/MemberController.java [hunk 11]
```diff
@@ -1,109 +1,108 @@
...
      * @param member
      * @return
      */
-    @RequestMapping(value = "/updateMember", method = {RequestMethod.GET, RequestMethod.POST})
+    @PutMapping
     @ResponseBody
-    public Map<String, Object> updateMember(@RequestBody Member member) {
-        System.out.println("member = " + member);
-        Map<String, Object> resultMap = new HashMap<>();
+    public Member updateMember(@RequestBody (required=false) Member member) {
```
Comment (f-lab-dori) [inline src/main/java/com/board/toyproject/controller/MemberController.java:77]: null체크보다 required = true로 하는 것이 낫지 않을까요?

File: src/main/java/com/board/toyproject/controller/MemberController.java [hunk 12]
```diff
@@ -1,109 +1,108 @@
...
+    @DeleteMapping
     @ResponseBody
-    public Map<String, Object> deleteMember(@RequestBody Member member) {
-        System.out.println("member = " + member);
-        String memberId=memberService.deleteMember(member);
-        Map<String, Object> resultMap = new HashMap<>();
-        resultMap.put("result", memberId+" deleted");
-        return resultMap;
+    public String deleteMember(@RequestBody (required=false) Member member) {
+        if (member == null) {
```
Comment (f-lab-dori) [inline src/main/java/com/board/toyproject/controller/MemberController.java:99]: 이것도 위에와 동일합니다.

File: src/main/java/com/board/toyproject/controller/exception/BadRequestException.java [hunk 13]
```diff
@@ -0,0 +1,8 @@
+package com.board.toyproject.controller.exception;
+
+public class BadRequestException extends RuntimeException {
+
+    public BadRequestException(String errorMsg) {
+        super(errorMsg);
```
Comment (f-lab-dori) [inline src/main/java/com/board/toyproject/controller/exception/BadRequestException.java:6]: Java에서는 변수명을 적을 때 단축어는 지양합니다.

File: src/test/java/com/board/toyproject/service/MemberServiceTest.java [hunk 14]
```diff
@@ -1,37 +1,80 @@
...
 
 @SpringBootTest
 @Transactional //테스트를 시작할 때 트랜잭션 시작하고 테스트 끝나면 롤백해줌.
 class MemberServiceTest {
 
     @Autowired MemberService memberService;
-    @Autowired MemberRepository memberRepository;
+
     @Test
     void 회원가입() {
```
Comment (f-lab-dori) [inline src/test/java/com/board/toyproject/service/MemberServiceTest.java:29]: method명은 영어로 해주세요. 그리고 edge case들을 추가해주세요

File: src/main/java/com/board/toyproject/controller/MemberController.java (hunk 8 above)
Comment (baeyuyeon) [reply src/main/java/com/board/toyproject/controller/MemberController.java:17]: 멘토님~ application.properties 10 line에
#uri prefix for all controller
server.servlet.context-path=/board/v1

로 추가하였는데 컨트롤러에 선언하는 것이 더 좋을 까요??
제가 다른 컨트롤러에도 같이 적용되도록 application.properties에 선언하였습니다!


그리고 url은 member로 수정하겠습니다.(멘토링시간에 복수형을 쓰면 좋은케이스를 여쭤보겠습니다!)

Comment (f-lab-dori) [reply src/main/java/com/board/toyproject/controller/MemberController.java:17]: 네 따로 하는 것이 좋아요... 왜냐하면 나중에 파일별로 버전을 분리할 수 있기 때문입니다.
